src/libopensc/pkcs15-esteid2018.c

```diff
@@ -0,0 +1,220 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "common/compat_strlcpy.h"
+
+#include "internal.h"
+#include "opensc.h"
+#include "pkcs15.h"
+
+static void set_string(char **strp, const char *value) {
+	if (*strp)
+		free(*strp);
+	*strp = value ? strdup(value) : NULL;
+}
+
+static int sc_pkcs15emu_esteid2018_init(sc_pkcs15_card_t *p15card) {
+	sc_card_t *card = p15card->card;
+	u8 buff[11];
+	int r, i;
+	size_t field_length = 0, taglen, j;
+	sc_path_t tmppath;
+
+	set_string(&p15card->tokeninfo->label, "ID-kaart");
+	set_string(&p15card->tokeninfo->manufacturer_id, "IDEMIA");
+
+	/* Read documber number to be used as serial */
+	sc_format_path("3F00D003", &tmppath);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, 11, 0);
+	LOG_TEST_RET(card->ctx, r, "read document number failed");
+	const unsigned char *tag = sc_asn1_find_tag(card->ctx, buff, (size_t)r, 0x04, &taglen);
+	if (tag == NULL)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	for (j = 0; j < taglen; j++)
+		if (!isalnum(tag[j]))
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	p15card->tokeninfo->serial_number = malloc(taglen + 1);
+	if (!p15card->tokeninfo->serial_number)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	p15card->tokeninfo->serial_number = memcpy(p15card->tokeninfo->serial_number, tag, taglen);
+	p15card->tokeninfo->serial_number[taglen] = '\0';
+	p15card->tokeninfo->flags = SC_PKCS15_TOKEN_READONLY;
+
+	/* add certificates */
+	for (i = 0; i < 2; i++) {
+		const char *esteid_cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+		const char *esteid_cert_paths[2] = {"3f00:adf1:3401", "3f00:adf2:341f"};
+		const u8 esteid_cert_ids[2] = {1, 2};
+
+		struct sc_pkcs15_cert_info cert_info;
+		struct sc_pkcs15_object cert_obj;
+
+		memset(&cert_info, 0, sizeof(cert_info));
+		memset(&cert_obj, 0, sizeof(cert_obj));
+
+		strlcpy(cert_obj.label, esteid_cert_names[i], sizeof(cert_obj.label));
+		sc_format_path(esteid_cert_paths[i], &cert_info.path);
+		cert_info.id.value[0] = esteid_cert_ids[i];
+		cert_info.id.len = 1;
+		r = sc_pkcs15emu_add_x509_cert(p15card, &cert_obj, &cert_info);
+		if (r < 0)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+		// Read data from first cert
+		if (i != 0)
+			continue;
+
+		sc_pkcs15_cert_t *cert = NULL;
+		r = sc_pkcs15_read_certificate(p15card, &cert_info, &cert);
```

Comment:
Fails to read Test card certificate. Reads only 1024 bytes.
Seems like the size is hardcoded here
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15.c#L2355

Reply:
Indeed. fixed this with latest push to this branch. But the right thing to do would be to remove that fixed 1k buffer from core so that the file could be read "until fully read" irrelevant of size.